### PR TITLE
Add skip-link to FSE themes

### LINKED ISSUE #21507: FSE: Add skip link
A skip link helps users skip repetitive blocks of content, like a header area and top navigation menu.

A skip link needs to:

- Be the first focusable object on the page
- Become visible when it receives focus
- Point to the content area of the site correctly

- A skip link may be hidden until it receives focus.
- Multiple skip links may also be used.

This is part of the WCAG guideline stating that a website must be navigable.
https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-skip.html
https://make.wordpress.org/themes/handbook/review/accessibility/required/#skip-links

Theme authors add this to their theme as the first element of the body manually, or
by using wp_body_open().

The target is usually the ID of the content wrapper, for example the main landmark.

Because this is an accessibility feature, it should be available for all themes,
not relying on the theme author to remember to add it.

Because the skip link must be the first element of a page, it is not suitable for a 
paragraph block with a link and a skip-link CSS class, because the block can then be moved and removed.

Here is one thought on how the skip link could work for FSE,
props @aristath:

_If the template in FSE contains a "content" block, then we could add an ID to it and automatically 
add the skip-link.
If there is no content block, then the skip-link would point to the loop block. 
If there is no loop-block either, then make an educated guess and point to the 1st title or p block._

### PR DESCRIPTION
**Pull request by aristath**
Fixes https://github.com/WordPress/gutenberg/issues/21507

## Description

This PR does the following:

* Adds a new `skipLinks` top-level item in `theme.json`
* Allows themes to opt-out of the auto-generated skip-link in theme.json
* Allows themes to define the skip-link target(s)
* Includes some sane fallback for elements that should be used as skip-link targets in case the author has not defined any targets, or the selector they defined was not found.

Then on render the following happen:

* We look inside the content HTML for things like `id="skip-link-target"` - assuming the user has defined `#skip-link-target` as the element in their theme.json file.
  * If the ID is located in the content, then we add the skip-link in `wp_body_open` and target the defined element.
  * If the ID is not located, then a vanilla-JS script gets injected in the footer and takes care of injecting the skip-link by querying the DOM for the fallback elements.

## `theme.json`

The `skipLinks` can be defined like this:
```json
{
	"skipLinks": {
		"auto": true,
		"css": true,
		"links": [
			{
				"target": "#skip-link-target",
				"label": "Skip to Content"
			},
			{
				"target": "#main-navigation",
				"label": "Skip to Navigation",
				"useFallbacks": false
			}
		]
	}
}
```
Arguments:
* `auto`: (bool) If set to false then automatic skip-links generation will be skipped. This is how authors can opt-out of the feature (for example if it's a single-page with no header and it goes straight to content :shrug:). If omitted then we assume `true`.
* `css:` (bool) We shouldn't force theme-authors to re-invent the wheel each time, so the CSS gets printed by default, but authors can opt-out if they want by setting it to `false`.
* `links`: (array) This is where users can define their skip-link (or multiple skip-links if needed). Each "link" can have the following:
  * `target`: (string|array) An element that should be used as the target.
  * `label`: (string) The skip-link's label. If omitted defaults to `Skip to content`.
  * `useFallbacks`: (bool) If set to false, then the fallbacks won't be added (more info about the fallbacks below). This only makes sense in case we have multiple skip-links defined.

## Fallback selectors
The fallback selectors are defined in an array:
```php
$fallbacks = array(
	'#skip-link-target',
	'main',
	'.wp-block-post-title',
	'.wp-block-query-loop',
	'.wp-block-post-content',
	'.entry-content',
	'h1',
	'h2',
);
```
This covers most (if not all) cases I could think of, and their priority makes sense but we can expand that logic and tweak them.

## Why not a separate block?

Adding a separate block for the skip-link and the skip-link target would introduce a lot of complexity for users. The UI would be cumbersome because it's basically invisible elements. Adding as a block would also allow people who don't necessarily understand the purpose or importance of a skip-link to point the skip-link to the wrong things (footer menus anyone?)
This automated implementation takes away all the guesswork, while still allowing users to point the skip-link to an element of their choosing if they want to, by adding the `skip-link-target` anchor to a block. It's a reasonable compromise between automation and freedom.

## How has this been tested?

Tested in an FSE theme:
* Index pages
* Archives
* Single posts
* Single pages

Tested both with a `#skip-link-target` element present, and without.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: lib/class-wp-theme-json.php
Comment (TimothyBJacobs) [inline lib/class-wp-theme-json.php]:
```suggestion
		return ! empty( $this->theme_json['skipLink']['css'] );
```

File: lib/full-site-editing/templates.php
Comment (TimothyBJacobs) [inline lib/full-site-editing/templates.php]: You could replace this with a `wp_json_encode` which might be a bit more readable.

File: lib/full-site-editing/templates.php
Comment (TimothyBJacobs) [inline lib/full-site-editing/templates.php]: I'm not sure if we care about this yet, but AFAIK using an anonymous function as a hook callback would be a violation of Core coding standards.

**Comment by gziolo** [comment]
An interesting idea 👍🏻 I have some questions related to the proposal. What if someone wants to have more than 1 skip link? What if they want to have a different text for the link? I'm a bit worried that the simplistic format that `theme.json` promotes creates limitations that might be hard to overcome. Just to give an example, in WP-Admin there are two skip links:

Screen Shot 2021-02-12 at 08 13 44
Screen Shot 2021-02-12 at 08 13 49

File: lib/full-site-editing/templates.php
Comment (aristath) [reply lib/full-site-editing/templates.php]: Good point. Fixed :+1:

File: lib/full-site-editing/templates.php
Comment (aristath) [reply lib/full-site-editing/templates.php]: Yeah, I was thinking about that too but the coding standard exists because we want users to be able to remove the hook.
In this case they can do it using `remove_action( 'wp_body_open', 'gutenberg_inject_skip_link' );` so I was thinking it's OK... I can see though why someone would want to remove just that one and not its parent.

Moved to a separate function in https://github.com/WordPress/gutenberg/pull/28946/commits/540a5cd108ec0ee9c62afeb7aac8a288527f3b71 :+1:

**Comment by nosolosw** [comment]
Hey, I was looking at this for the code changes, but I believe is better that we clarify direction before diving into code.

According to https://github.com/WordPress/gutenberg/issues/21507#issuecomment-730703562 it looks like the direction we want to go is trying to absorb the concept of skip links into the template parts (header). I lack context on the specifics but I understand it can be a block or an attribute of the template part ― for which we need the ability to "understand" what's a header, which is under development judging by the status of https://github.com/WordPress/gutenberg/issues/26599 although a first PR to classify them has recently landed https://github.com/WordPress/gutenberg/pull/28410

**Comment by carolinan** [comment]
The skip link needs to be the first focusable item on a page, no matter where there or if a header template part is used.

**Comment by aristath** [comment]
> What if someone wants to have more than 1 skip link?

Interesting and valid! I refactored the implementation to allow for multiple skip-links and adjusted the PR description accordingly :+1:

**Comment by carolinan** [comment]
Please include the documentation for the new feature.

**Comment by aristath** [comment]
> Please include the documentation for the new feature.

If it gets to the point where it is agreed that we want to merge this then of course docs will be written. But not before then, it would be a waste of time :+1:

**Review by carolinan — APPROVED** [review]

**Comment by carolinan** [comment]
Is there anything still blocking this?  I guess it needs two approvals?

**Comment by aristath** [comment]
Added docs.

**Comment by aristath** [comment]
cc @JustinyAhin for the docs in this PR 👍

**Comment by JustinyAhin** [comment]
> cc @JustinyAhin for the docs in this PR 👍

Docs look good to me.
Thank you for writing them :)

**Comment by carolinan** [comment]
I believe the docs covers everything.

**Comment by aristath** [comment]
Though approved, I'd like some more feedback here before merging a change like this. It's important we get it right. 👍 
I'll leave this open for a while, give everyone the opportunity to voice their opinion.

File: lib/full-site-editing/templates.php
Comment (gziolo) [inline lib/full-site-editing/templates.php:207]: Isn't `screen-reader-text` already defined in WordPress or is it only for WP-Admin? See: https://make.wordpress.org/accessibility/2015/02/09/hiding-text-for-screen-readers-with-wordpress-core/

Comment (aristath) [reply lib/full-site-editing/templates.php:207]: Unfortunately, that's only for wp-admin... All themes add that same CSS manually in their CSS files. 🤷

**Comment by aristath** [comment]
Converted to draft for now, after some discussion #30336 is simpler and more likely to land at this stage.

**Comment by aristath** [comment]
Closing this one, #30336 was merged instead.

**Pull request closed without merging**
